torchtitan/models/llama/model.py

```diff
@@ -208,13 +224,34 @@ def forward(
         xv = values.transpose(1, 2)  # (bs, n_local_heads, seqlen, head_dim)
 
         # we use casual mask for training
-        output = F.scaled_dot_product_attention(xq, xk, xv, is_causal=True)
+        if self.use_flex_attn:
+            # assert False, (type(xq), type(xk), type(xv))
+            self._init_flex_attn(seqlen=seqlen)
+            output = self.flex_attn(xq, xk, xv, block_mask=self.block_mask)
+        else:
+            output = F.scaled_dot_product_attention(xq, xk, xv, is_causal=True)
         output = output.transpose(
             1, 2
         ).contiguous()  # (bs, seqlen, n_local_heads, head_dim)
         output = output.view(bs, seqlen, -1)
         return self.wo(output)
 
+    @torch.no_grad()
+    def _init_flex_attn(self, seqlen: int) -> None:
```

Comment:
IMO maybe making a SDPA module that switches between fsdp and flex attention might be better instead of embedding them to each model files